modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteSnapshotManager.java

```diff
@@ -3775,57 +3775,42 @@ private synchronized void scheduleNext() {
         }
 
         /** Stopping handler. */
-        public synchronized void stop() {
-            stopping = true;
-
-            if (active != null)
-                active.acceptException(new IgniteException(SNP_NODE_STOPPING_ERR_MSG));
+        public void stop() {
+            synchronized (this) {
+                stopping = true;
+            }
 
             RemoteSnapshotFilesRecevier r;
 
             while ((r = queue.poll()) != null)
                 r.acceptException(new IgniteException(SNP_NODE_STOPPING_ERR_MSG));
 
-            Set<RemoteSnapshotFilesRecevier> futs = activeTasks();
-            GridCompoundFuture<Void, Void> stopFut = new GridCompoundFuture<>();
-
-            try {
-                for (IgniteInternalFuture<Void> fut : futs)
-                    stopFut.add(fut);
-
-                stopFut.markInitialized().get();
-            }
-            catch (IgniteCheckedException e) {
-                throw new IgniteException(e);
-            }
+            if (active != null)
+                active.acceptException(new IgniteException(SNP_NODE_STOPPING_ERR_MSG));
         }
 
-        /**
-         * @param nodeId A node left the cluster.
-         */
+        /** @param nodeId A node left the cluster. */
         public void onNodeLeft(UUID nodeId) {
-            Set<RemoteSnapshotFilesRecevier> futs = activeTasks();
             ClusterTopologyCheckedException ex = new ClusterTopologyCheckedException("The node from which a snapshot has been " +
                 "requested left the grid");
 
-            futs.forEach(t -> {
-                if (t.rmtNodeId.equals(nodeId))
-                    t.acceptException(ex);
+            queue.forEach(r -> {
+                if (r.rmtNodeId.equals(nodeId))
+                    r.acceptException(ex);
             });
-        }
-
-        /**
-         * @return The set of currently scheduled tasks, some of them may be already completed.
-         */
-        private Set<RemoteSnapshotFilesRecevier> activeTasks() {
-            Set<RemoteSnapshotFilesRecevier> futs = new HashSet<>(queue);
 
             RemoteSnapshotFilesRecevier active0 = active;
 
-            if (active0 != null)
-                futs.add(active0);
-
-            return futs;
+            if (active0 != null && !active0.isDone() && active0.rmtNodeId.equals(nodeId)) {
```

Comment:
I'm not sure that a waiting is a good reason. Can we just stop every transmission if a required node leaves?

Reply:
Most tasks have already stopped by this point. This is for the rare case when the node has successfully send `SnapshotFilesRequestMessage` and waiting for transmission BUT not received `SnapshotFilesFailureMessage` message (and transmission was not started) .